Bjorncs/jrt ssl socket

bjorncs opened the pull request.

bjorncs committed:
Add ssl socket implementation based on SSLEngine

bjorncs committed:
Test SslEngine in EchoTest and SessionTest

havardpe commented on jrt/src/com/yahoo/jrt/SslSocket.java:
```diff
@@ -0,0 +1,232 @@
+ *
+ * @author bjorncs
+ */
+public class SslSocket implements CryptoSocket {
```
I think SslSocket is too close to SSLSocket, I suggest TlsCryptoSocket instead

havardpe commented on jrt/src/com/yahoo/jrt/SslSocket.java:
```diff
@@ -0,0 +1,232 @@
+    private final SSLEngine sslEngine;
+    private final ByteBuffer wrapBuffer;
+    private final ByteBuffer unwrapBuffer;
+    private final ByteBuffer handshakeDummyBuffer;
```
consider discarding this after handshake has completed

havardpe commented on jrt/src/com/yahoo/jrt/SslSocket.java:
```diff
@@ -0,0 +1,232 @@
+    private int sslEngineAppDataUnwrap(ByteBuffer dst) throws IOException {
+        int bytesUnwrapped = sslEngineUnwrap(dst);
+        if (bytesUnwrapped == 0) throw new SSLException("Got handshake data in application data unwrap");
+        if (bytesUnwrapped == -1) return 0;
```
seems a bit inconsistent to map -1 to 0 in unwrap, but not in wrap (for app data)

havardpe commented on jrt/src/com/yahoo/jrt/SslSocket.java:
```diff
@@ -0,0 +1,232 @@
+
+    @Override
+    public int write(ByteBuffer src) throws IOException {
+        channelWrite();
```
consider calling flush method here

havardpe commented on jrt/src/com/yahoo/jrt/SslSocket.java:
```diff
@@ -0,0 +1,232 @@
+
+    private final SocketChannel channel;
+    private final SSLEngine sslEngine;
+    private final ByteBuffer wrapBuffer;
```
consider using jrt.Buffer class for buffers to not need to keep track of read/write mode and content size/location (for future consideration, not blocking this PR)

bjorncs replied:
I have changed `wrapBuffer` and `unwrapBuffer` to use `com.yahoo.jrt.Buffer` in latest commit

havardpe reviewed (commented):
looks good overall, please consider my comments

vekterli commented on jrt/src/com/yahoo/jrt/SslSocket.java:
```diff
@@ -0,0 +1,232 @@
+                case NEED_UNWRAP:
+                    if (hasWrapBufferMoreData()) return HandshakeState.NEED_WRITE;
+                    int bytesUnwrapped = sslEngineUnwrap(handshakeDummyBuffer);
+                    if (handshakeDummyBuffer.position() > 0) throw new SSLException("Got application data in handshake unwrap: " + handshakeDummyBuffer);
```
Was pondering if this could trigger if a server peer pipelines its `Finished` packet with the first application data, but assuming unwrap() only processes 1 handshake frame per invocation during handshaking it _shouldn't_. Are my assumptions correct?

bjorncs replied:
Only a single frame is unwrapped at a time.

bjorncs committed:
Various fixes to SslSocket

- Clear reference to handshake dummy buffer after completed handshake.
- Make return value consistent between sslEngineAppDataWrap and sslEngineAppDataUnwrap.
- Throw SSLException in sslEngineUnwrap when destination buffer capacity is too low.
- SocketChannel.write() never returns -1
- Call flush() from write()

bjorncs committed:
Rename SslEngine/SslSocket -> TlsCryptoEngine/TlsCryptoSocket

vekterli reviewed (approved):
:+1: just have a small question re: SSLEngine semantics

havardpe commented on jrt/src/com/yahoo/jrt/SslSocket.java:
```diff
@@ -179,6 +181,8 @@ private int sslEngineUnwrap(ByteBuffer dst) throws IOException {
             case OK:
                 return result.bytesProduced();
+            case BUFFER_OVERFLOW:
+                throw new SSLException("Cannot unwrap - remaining capacity too small: " + dst);
```
You cannot do this as long as you have a loop around the call to unwrap. A channel read could read 2 frames (since there is extra room in the input buffer) and the output buffer could be too small even if it adheres to the buffer size requirements.

bjorncs replied:
Good catch - it should return 0 instead.

bjorncs replied:
I have updated unwrap logic to use separate code path for handling overflow. Overflow will no longer throw exception.

havardpe commented on jrt/src/com/yahoo/jrt/SslSocket.java:
```diff
@@ -147,12 +148,12 @@ private int sslEngineAppDataUnwrap(ByteBuffer dst) throws IOException {
     public int write(ByteBuffer src) throws IOException {
-        channelWrite();
-        if (hasWrapBufferMoreData()) return 0;
+        FlushResult flushResult = flush();
```
no need to capture the flush return value since it is only used in the if below

bjorncs committed:
Rewrite wrap+unwrap to remove use of magic return values

- Wrap/unwrap for handshake returns true for success, false otherwise
- Wrap/unwrap for application data returns bytes consumed/produced
- Do not throw exception on overflow for unwrap
- Misc changes to reduce code duplication

bjorncs committed:
Use com.yahoo.jrt.Buffer to simplify buffer management

havardpe commented on jrt/src/com/yahoo/jrt/TlsCryptoSocket.java:
```diff
@@ -0,0 +1,254 @@
+    public int write(ByteBuffer src) throws IOException {
+        if (flush() == FlushResult.NEED_WRITE) return 0;
+        int totalBytesWrapped = 0;
+        while (src.hasRemaining()) {
```
why is this loop different from the same loop in the other direction?

havardpe commented on jrt/src/com/yahoo/jrt/TlsCryptoSocket.java:
```diff
@@ -0,0 +1,254 @@
+            throw new SSLException("Handshake not completed: handshakeState=" + handshakeState);
+    }
+
+    private boolean hasWrapBufferMoreData() {
```
might be just as understandable to just ask the buffer directly?

havardpe commented on jrt/src/com/yahoo/jrt/TlsCryptoSocket.java:
```diff
@@ -0,0 +1,254 @@
+    }
+
+    private SSLEngineResult sslEngineWrap(ByteBuffer src) throws IOException {
+        SSLEngineResult result = sslEngine.wrap(src, wrapBuffer.getWritable(0));
```
should give network frame size as parameter instead of 0 here

havardpe commented on jrt/src/com/yahoo/jrt/TlsCryptoSocket.java:
```diff
@@ -0,0 +1,254 @@
+
+    // returns number of bytes read
+    private int channelRead() throws IOException {
+        int read = channel.read(unwrapBuffer.getWritable(0));
```
should give something larger than 0 as parameter to ensure we try to read a reasonable chunk of data

havardpe commented on jrt/src/com/yahoo/jrt/TlsCryptoSocket.java:
```diff
@@ -0,0 +1,253 @@
+    @Override
+    public int read(ByteBuffer dst) throws IOException {
+        verifyHandshakeCompleted();
+        int bytesUnwrapped = applicationDataUnwrap(dst);
```
consider calling drain here to avoid returning a single frame even if multiple frames could be decoded

havardpe commented:
looks good, minor comment

bjorncs committed:
Use drain() to unwrap multiple frames before reading from network
